### DISCUSSION TITLE
Cannot get PyOpenCL installed *with GL sharing* on Linux

### DISCUSSION THREAD

**Discussion by AwesomeCronk** [pyopencl/pyopencl#554] [Troubleshooting] [answered]
I see this has replaced the mailing list, so I shall try here to get help.

I tried installing PyOpenCL with GL support on my laptop with an Intel iGPU and it didn't work, so I chalked it up as a driver issue. Well now I'm trying it on my desktop with a Nvidia GPU and it's still not working.
To be clear, the Linux wheels *do work.* If I just `python3 -m pip install pyopencl` I get a working PyOpenCL install and I can do whatever *except GL interops!*
Now, I want to visualize the results of running my CL programs directly, without running the data back to Python and painting on a QtWidget, because that's very slow. So I tried installing PyOpenCL from the release tarball at the PyPI project page. Here's how it went:
1. `python3 -m pip uninstall pyopencl` - remove the wheel version
2. Unpack the source release
3. `cd pyopencl-2022.1`
4. `python3 configure.py --cl-enable-gl` - This is supposed to enable GL interops.
5. `python3 -m pip install .` - fails, so I run `sudo apt install ocl-icd-opencl-dev` and rerun this pip command.
6. Now I cd to the folder where my Python program (attached here) is and `python3 PyOpenCL_GL.py`:
```python
Traceback (most recent call last):
  File "PyOpenCL_GL.py", line 9, in <module>
    import pyopencl as cl
  File "/home/aweso/.local/lib/python3.8/site-packages/pyopencl/__init__.py", line 28, in <module>
    import pyopencl.cltypes  # noqa: F401
  File "/home/aweso/.local/lib/python3.8/site-packages/pyopencl/cltypes.py", line 22, in <module>
    from pyopencl.tools import get_or_register_dtype
  File "/home/aweso/.local/lib/python3.8/site-packages/pyopencl/tools.py", line 36, in <module>
    from pyopencl._cl import bitlog2  # noqa: F401
ImportError: /home/aweso/.local/lib/python3.8/site-packages/pyopencl/_cl.cpython-38-x86_64-linux-gnu.so: undefined symbol: clSetProgramSpecializationConstant
```
8. At this point I get very frustrated and confused and run `clinfo` to see just what the heck is up (see attached clinfo_dump.txt). I also run `cat /etc/OpenCL/vendors/nvidia.icd` and get `libnvidia-opencl.so.1` as a result. `which libnvidia-opencl.so.1` yields nothing.

I am on Ubuntu Linux 20.04 LTS., This machine has a Core i5-7600k and MSI/Nvidia GTX1650. If anyone can point me to solving this, I would greatly appreciate it!

[clinfo_dump.txt](https://github.com/inducer/pyopencl/files/8455742/clinfo_dump.txt)
[PyOpenCL_GL.py.txt](https://github.com/inducer/pyopencl/files/8455744/PyOpenCL_GL.py.txt)

**Comment by inducer** [comment, marked as answer]
Try the following:

Add `--cl-pretend-version=2.1` to your `configure.py` line. Run `rm -Rf build` to ensure the package actually gets recompiled. Reinstall.

This is needed because (at least at one point) ocl-icd included headers declaring CL 2.2 functions (like `clSetProgramSpecializationConstant`) but did not actually define them.

**Reply by AwesomeCronk** [reply]
Hot dang, thank you! Now I can look into getting it set up on my Intel machine as well. Thank you ever so much!

Edit: Haha, nope! Will not work on an Intel iGPU with Linux and NEO until it supports cl_khr_gl_sharing. Works on Nvidia tho :man_shrugging:
